### ISSUE TITLE
[R] Low accuracy in CIFAR with R and resnet

### ISSUE THREAD

**Issue by hoaphumanoid** [dmlc/mxnet#3097]
I computed Resnet network in R, this is the code:

```R
get_iterator <- function(data.shape) {
    data_dir = args$data_dir
    data.shape <- data.shape
    train = mx.io.ImageRecordIter(
      path.imgrec     = paste0(data_dir, "train.rec"),
      path.imglist    = paste0(data_dir, "train.lst"),
      batch.size      = args$batch_size,
      data.shape      = data.shape,
      shuffle         = TRUE
    )  
    val = mx.io.ImageRecordIter(
      path.imgrec     = paste0(data_dir, "test.rec"),
      path.imglist    = paste0(data_dir, "test.lst"),
      batch.size      = args$batch_size,
      data.shape      = data.shape
    )
    ret = list(train=train, value=val)
}
parse_args <- function() {
  parser <- ArgumentParser(description='train an image classifer on CIFAR')
  parser$add_argument('--network', type='character', default='resnet-28-small',
                      choices = c('mlp', 'lenet'),
                      help = 'the cnn to use')
  parser$add_argument('--data-dir', type='character', default='../../data/cifar10/',
                      help='the input data directory')
  parser$add_argument('--gpus', type='character',
                      help='the gpus will be used, e.g "0,1,2,3"')
  parser$add_argument('--batch-size', type='integer', default=128,
                      help='the batch size')
  parser$add_argument('--lr', type='double', default=.1,
                      help='the initial learning rate')
  parser$add_argument('--model-prefix', type='character',
                      help='the prefix of the model to load/save')
  parser$add_argument('--num-round', type='integer', default=10,
                      help='the number of iterations over training data to train the model')
  parser$add_argument('--kv-store', type='character', default='local',
                      help='the kvstore type')
  parser$parse_args()
}
args = parse_args()

# train
if (args$network == 'resnet') {
  source("symbol_resnet.R")
  net <- get_symbol()
} else {
  source("symbol_resnet-28-small.R")
  net <- get_symbol()
}

# save model
if (is.null(args$model_prefix)) {
  checkpoint <- NULL
} else {
  checkpoint <- mx.callback.save.checkpoint(args$model_prefix)
}

# data
data.shape <- c(28,28,3)
data <- get_iterator(data.shape = data.shape)
train <- data$train
val <- data$value


# train
if (is.null(args$gpus)) {
  print("Computing with CPU")
  devs <- mx.cpu()  
} else {
  print(paste0("GPU option: ", args$gpus))
  devs <- lapply(unlist(strsplit(args$gpus, ",")), function(i) {
    mx.gpu(as.integer(i))
  })
}

#train
time_init <- Sys.time()
model = mx.model.FeedForward.create(
  X                  = train,
  eval.data          = val,
  ctx                = devs,
  symbol             = net,
  eval.metric        = mx.metric.accuracy,
  num.round          = args$num_round,
  learning.rate      = args$lr,
  momentum           = 0.9,
  wd                 = 0.00001,
  kvstore            = args$kv_store,
  array.batch.size   = args$batch_size,
  epoch.end.callback = checkpoint,
  batch.end.callback = mx.callback.log.train.metric(50)
)
time_end <- Sys.time()
difftime(time_end, time_init, units = "mins")

```
I got a very low accuracy:

```R
Start training with 1 devices
Batch [50] Train-accuracy=0.10296875
Batch [100] Train-accuracy=0.099609375
Batch [150] Train-accuracy=0.0980208333333333
Batch [200] Train-accuracy=0.0970703125
Batch [250] Train-accuracy=0.0973125
Batch [300] Train-accuracy=0.098125
Batch [350] Train-accuracy=0.0989508928571429
[1] Train-accuracy=0.103485576923077
[1] Validation-accuracy=0.181170886075949
Batch [50] Train-accuracy=0.09609375
...
...
Batch [50] Train-accuracy=0.10625
Batch [100] Train-accuracy=0.100078125
Batch [150] Train-accuracy=0.101041666666667
Batch [200] Train-accuracy=0.101953125
Batch [250] Train-accuracy=0.1025
Batch [300] Train-accuracy=0.107760416666667
Batch [350] Train-accuracy=0.116964285714286
[10] Train-accuracy=0.128996163682864
[10] Validation-accuracy=0.259915865384615
```

Any idea?

**Comment by winstywang** [comment]
Try warm up with a smaller lr as stated in the paper.

**Comment by maxkazmsft** [comment]
We tried warming up with 0.01 learning rate like in the paper, and even 0.005 learning rate. There appears to be a large discrepancy between R and Python implementations. Python implementation works very well, but in R there appears to be something wrong with how the error on the validation set is computed - please see the attached plot. Even with 0.005 learning rate, it makes no sense for R implementation to have better accuracy on the validation set than on the training set in the first 60 epochs with the default settings. The validation set accuracy is also extremely volatile.
![accuracy](https://cloud.githubusercontent.com/assets/19535372/17980070/abfcbe02-6acb-11e6-9c44-a8aae050e8f7.PNG)

**Comment by maxkazmsft** [comment]
Looks like we will need to adjust the input data - there seems to be a mismatch between how the data is fed into Python and R.
